### PR TITLE
Support tuple conversions in Expression.Convert and ConvertChecked

### LINKED ISSUE #18284: Expression.Convert and Expression.ConvertChecked should support tuple conversions.
Currently the code:
```C#
Expression.Convert(Expression.Constant((1, 1)), typeof((long, long)))
```
Throws an `InvalidOperationException` with the message
```C#
"No coercion operator is defined between types 'System.ValueTuple`2[System.Int32,System.Int32]' and 'System.ValueTuple`2[System.Int64,System.Int64]'."
```

As such conversions are now among the standard conversions one expects to work with the main .NET languages, they should work with S.L.Expressions too.

As well as being of value in itself, this will facilitate #17453 as then MS.CSharp can be changed to recognise when such conversions are relevant and create the relevant Linq expression, and would be at least beneficial to dotnet/roslyn#12897

### PR DESCRIPTION
**Pull request by JonHanna**
Closes #18284

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs [hunk 1]
```diff
@@ -246,6 +246,70 @@ public static bool HasIdentityPrimitiveOrNullableConversionTo(this Type source,
...
+                case 4:
+                    return type == typeof(ValueTuple<,,,>);
+                case 5:
+                    return type == typeof(ValueTuple<,,,,>);
+                case 6:
+                    return type == typeof(ValueTuple<,,,,,>);
+                case 7:
+                    return type == typeof(ValueTuple<,,,,,,>);
+                case 8:
+                    return type == typeof(ValueTuple<,,,,,,,>);
```
Comment (OmarTawfik) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: Perhaps take a different action when having a different `argumentCount` than returning false?

Comment (OmarTawfik) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: TOL if it should just be an assert...

---
In reply to: [112053001](https://github.com/dotnet/corefx/pull/18289#discussion_r112053001) [](ancestors = 112053001)

**Comment by VSadov** [comment]
Need to think about this a bit. This is to support a new feature. 

Evidently we have collected some debt around new language features and their use in dynamic. There is a concern though that we should have an overall strategy for the fixes - how, when, where and what will be fixed. For example what it would mean for other implementations (line NET FX) , whether we want new nodes or just reduce to the existing set, and so on.

We have other issues around tuples and dynamic, such as 
https://github.com/dotnet/corefx/issues/17453
https://github.com/dotnet/corefx/issues/17225

I will keep this open for now, but I think we will need to answer more design questions before embarking on supporting tuples in ET and dynamic.

File: src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs]: A nonary or higher generic type could legitimately come here, and false would be the correct answer.

**Comment by karelz** [comment]
Given where, we are, I would suggest to block this merge until after 5/10 - see https://github.com/dotnet/corefx/issues/17619#issuecomment-296872132

**Comment by danmosemsft** [comment]
Master open again for htis kind of change.
@dotnet-bot test this please.

**Comment by JonHanna** [comment]
Test Innerloop netfx Windows_NT Debug x86 Build and Test  please.

**Comment by VSadov** [comment]
Do not merge this just yet. 
- Need to decide whether we need a special conversion kind for tuple conversions 
- how this relates to all other technical debt WRT new C# language features.
- since this is a new functionality. Do we need to bump the version number.

**Comment by karelz** [comment]
@VSadov what's the ETA for next step?

**Comment by VSadov** [comment]
I need more time to get to this. 
I have added the matter of the technical debt WRT Linq.Expressions/dynamic and the new (v6.0+) language features to the list of items to discuss at the language design. 
Is not on the top of the list though, since there are more pressing issues. We simply did not get to this yet.

**Comment by karelz** [comment]
OK, given that this is 2 month old PR, do you have rough ETA?
Or is it blocked by language design discussion? (that feels weird, but your reply kind of hinted that)

While we have an excuse of shipping .NET Core 2.0 in the last 2 months (hence being super-busy), I want to make sure we do not postpone PR reviews too much without really good reasons. Thanks!

**Comment by VSadov** [comment]
Yes. This is blocked on the language discussion. We need to have a sense on the direction of the language design in this area.
I do not see this as an isolated change/fix related to tuple conversions. We have a number of other new language features not supported in the same fashion. 

I have passed this item to the LDM, but the recent design sessions were dominated by C#7.2 issues.

**Comment by karelz** [comment]
@VSadov I would recommend to close this PR in such case. We do not want to have stale PRs which don't have clear way forward.
If you can point @JonHanna to the C# discussions / bugs filed, that would be great. He may want to participate ...

**Pull request closed without merging**